### PR TITLE
Fix: For cached/pre-computed change points, disabled/enabled metric i…

### PR DESCRIPTION
**Pull request by henrikingo**
…nvalidates cache

### CODE REVIEW

**Comment by henrikingo** [comment]
@mfleming Ok that was messier than it should have...

I'm not particularly proud of last_modified now leaking outside the API to clients. That said it could be useful in client side caching or something completely dfferent... better sleep on it.

Either way it's harmless and this is a good enough patch to unblock other merges. Can change it or leave it later.

**Comment by mfleming** [comment]
> @mfleming Ok that was messier than it should have...
> 
> I'm not particularly proud of last_modified now leaking outside the API to clients. That said it could be useful in client side caching or something completely dfferent... better sleep on it.
> 
> Either way it's harmless and this is a good enough patch to unblock other merges. Can change it or leave it later.

Could we add `last_modified` to `user.py::validate_config()` to prevent users from writing it, i.e. make it read-only.

**Review by mfleming — APPROVED** [review]

**Comment by henrikingo** [comment]
> Could we add last_modified to user.py::validate_config() to prevent users from writing it, i.e. make it read-only.

Certainly and good suggestion. I'll do it later today.

**Comment by henrikingo** [comment]
Do you want me to merge this asap or do you want to still discuss alternatives / improvements?

What I was thinking as an alternative to the current PR, is that db.py would be responsible for handling last_modified, but would return it separately when required. Something like:

    doc = db.coll.find({_id:pk}
    last_modified = doc["last_modified"]
    del doc["last_modified"]
    return doc, {"last_modified": last_modified}

This would contain the responsibility to not leak last_modified outside the backend API to a single class and file. Callers to db.py methods would still be exposed to this, but could just

    doc, _ = store.get_foo(id)

...in cases they don't need it.

**Comment by mfleming** [comment]
Yeah plus do the alternative that sounds better

File: backend/api/api.py [hunk 1]
```diff
@@ -131,7 +129,12 @@ async def get_result(
     if not list(filter(lambda name: name == test_name, test_names)):
         raise HTTPException(status_code=404, detail="Not Found")
 
-    return _strip_last_modified(await store.get_results(user.id, test_name))
+    results = await store.get_results(user.id, test_name)
+    print(results)
+    if not results:
+        return [{}]
```
Comment (henrikingo) [inline backend/api/api.py:135]: Enforced by pydantic and I didn't want to touch the return values so playing along... (But it's silly and needs to change eventually.)

**Comment by henrikingo** [comment]
@mfleming Here a patch with alternative 2. Some pydantic validations still failing. It's a mystery to me how I'm supposed to fix them, when it only has a stack trace of its own methods leading to the failure, but nothing hinting at where  the problem might be in nyrkio code.

I spent most of the day fighting pydantic, need a break now.

**Comment by henrikingo** [comment]
Ok so the patch doesn't look that terrible right now, it was just a long way back and forth to get here.

The fundamental question is, when a method in db.py returns a list  of results, what do we prefer?

    return data, meta
    
    # where...
    data = [{json:doc1}, {json:doc2}, ...] 
    meta = [meta1, meta2...]
    
Or...

   return list_of_tuples
    
    # where
    list_of_tuples = [(data1, meta1), (data2, meta2),...]

The benefit of the first alternative is that when you don't need the ´meta["lastmodified"]´ info,which is most cases, you can just:

    data, _ = store.get_something()

...and you're good to go. Same code works without changes.

The downside is if you do need last_modified info, how do you iterate over both lists together? I believe

    for d,m in zip(data,meta):
        do_something(d["user_id"], d["something else"], m["last_modified"])

...is the right way in python?

Conversely, with the second approach iterating over all data objects, it's actully quite convenient but not at all intuitive to just:
    
    for obj in results:
        do_something(obj(0))

**Comment by mfleming** [comment]
I guess the question is which use is more common and the pick type for that. I'm fine either way (yes zip is the correct way in Python).

**Comment by henrikingo** [comment]
Ok, I'll try switching to the other way. Partly thanks to zip it feels more natural now.

**Pull request merged**